testkit/tests/supervisor/main.rs

```diff
@@ -0,0 +1,437 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use exonum_testkit::{ApiKind, InstanceCollection, TestKit, TestKitApi, TestKitBuilder};
+use serde_json::json;
+
+use exonum::{
+    api::{
+        self,
+        node::public::{explorer::TransactionQuery, system::DispatcherInfo},
+    },
+    crypto::{self, Hash},
+    helpers::Height,
+    proto::Any,
+    runtime::{
+        rust::Transaction,
+        supervisor::{DeployRequest, StartService},
+        ArtifactId,
+    },
+};
+
+use crate::inc::{IncService, TxInc, SERVICE_ID, SERVICE_NAME};
+
+mod inc;
+mod proto;
+
+const RUNTIME_ID_RUST: u32 = 0;
+const ARTIFACT_NAME: &str = "inc/1.0.0";
+
+fn artifact_default() -> ArtifactId {
+    ArtifactId {
+        runtime_id: RUNTIME_ID_RUST,
+        name: ARTIFACT_NAME.into(),
+    }
+}
+
+fn assert_count(api: &TestKitApi, service_name: &'static str, expected_count: u64) {
+    let real_count: u64 = api
+        .public(ApiKind::Service(service_name))
+        .get("v1/counter")
+        .unwrap();
+    assert_eq!(real_count, expected_count);
+}
+
+fn assert_no_count(api: &TestKitApi, service_name: &'static str) {
+    let response: api::Result<u64> = api.public(ApiKind::Service(service_name)).get("v1/counter");
+    assert!(response.is_err());
+}
+
+fn assert_tx_status(api: &TestKitApi, tx_hash: Hash, expected_status: &serde_json::Value) {
+    let info: serde_json::Value = api
+        .public(ApiKind::Explorer)
+        .query(&TransactionQuery::new(tx_hash))
+        .get("v1/transactions")
+        .unwrap();
+    if let serde_json::Value::Object(mut info) = info {
+        let tx_status = info.remove("status").unwrap();
+        assert_eq!(tx_status, *expected_status);
+    } else {
+        panic!("Invalid transaction info format, object expected");
+    }
+}
+
+fn dispatcher_info(api: &TestKitApi) -> DispatcherInfo {
+    api.public(ApiKind::System).get("v1/services").unwrap()
+}
+
+fn find_instance_id(api: &TestKitApi, instance_name: &'static str) -> u32 {
+    let info = dispatcher_info(api);
+    info.services
+        .iter()
+        .find(|service| service.name == instance_name)
+        .expect("Can't find the instance")
+        .id
+}
+
+fn deploy_artifact(api: &TestKitApi, request: DeployRequest) -> crypto::Hash {
+    let hash: crypto::Hash = api
+        .private(ApiKind::Service("supervisor"))
+        .query(&request)
+        .post("deploy-artifact")
+        .unwrap();
+    hash
+}
+
+fn start_service(api: &TestKitApi, request: StartService) -> crypto::Hash {
```

Comment:
You have trait for supervisor `PrivateApi` https://github.com/exonum/exonum/blob/dynamic_services/exonum/src/runtime/supervisor/api.rs#L28
Just implement it for `TestKitApi`

Reply:
It's possible, but it seems to require even more boilerplate code (see commented-out code in the current PR draft). Is it worth it?